File: Ravi-Teja/LetterCasePermutation.cpp
class Solution {
public:
    void dfs(int n, string &s, vector<string> &result){
        if(n<0){
            result.push_back(s);
            return;
        }
        dfs(n-1, s, result);
        if(s[n]>='a' && s[n]<='z'){
            s[n]=(s[n]-'a')+'A';
        }
        else if(s[n]>='A' && s[n]<='Z'){
            s[n]=(s[n]-'A')+'a';
        }
        else{
            return;
        }
        dfs(n-1, s, result);
    }
    vector<string> letterCasePermutation(string s) {
        vector<string> result;
        dfs(s.length()-1, s, result);
        return result;
    }
};

File: Parth Mathur/letterCasePermutation.java
class Solution {
    public List<String> letterCasePermutation(String s) {
        List<String> result=new ArrayList<>();
        backtrack(result, s, new StringBuilder(), 0);
        return result;
    }
    public void backtrack(List<String> result, String s, StringBuilder sb, int ind){
        if(ind==s.length()){
            StringBuilder copy=new StringBuilder(sb);
            result.add(copy.toString());
            return;
        }
        char curr=s.charAt(ind);
        if(Character.isDigit(curr)){
            sb.append(curr);
            backtrack(result, s, sb, ind+1);
            sb.deleteCharAt(sb.length()-1);
        }else{
            sb.append(Character.toLowerCase(curr));
            backtrack(result, s, sb, ind+1);
            sb.deleteCharAt(sb.length()-1);
            sb.append(Character.toUpperCase(curr));
            backtrack(result, s, sb, ind+1);
            sb.deleteCharAt(sb.length()-1);
        }
    }
}
